pkg/cloudprovider/metrics/cloudprovider.go

```diff
@@ -0,0 +1,143 @@
+/*
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package metrics
+
+import (
+	"context"
+	"time"
+
+	"github.com/aws/karpenter/pkg/apis/provisioning/v1alpha5"
+	"github.com/aws/karpenter/pkg/cloudprovider"
+	"github.com/aws/karpenter/pkg/metrics"
+	"github.com/aws/karpenter/pkg/utils/injection"
+	"github.com/prometheus/client_golang/prometheus"
+	v1 "k8s.io/api/core/v1"
+	"knative.dev/pkg/apis"
+	"knative.dev/pkg/logging"
+	crmetrics "sigs.k8s.io/controller-runtime/pkg/metrics"
+)
+
+const (
+	metricLabelComponent = "component"
+	metricLabelMethod    = "method"
+	metricLabelProvider  = "provider"
+	metricLabelResult    = "result"
+)
+
+var methodDurationHistogramVec = prometheus.NewHistogramVec(
+	prometheus.HistogramOpts{
+		Namespace: metrics.Namespace,
+		Subsystem: "cloudprovider",
+		Name:      "duration_seconds",
+		Help:      "Duration of cloud provider method calls.",
+	},
+	[]string{
+		metricLabelComponent,
+		metricLabelMethod,
+		metricLabelProvider,
+		metricLabelResult,
+	},
+)
+
+func init() {
+	crmetrics.Registry.MustRegister(methodDurationHistogramVec)
+}
+
+type decorator struct {
+	cloudprovider.CloudProvider
+}
+
+// PublishLatency returns a new `CloudProvider` instance that will delegate all method
+// calls to the argument, `cloudProvider`, and publish aggregated latency metrics. The
+// value used for the metric label, "component", is taken from the `Context` object
+// passed to the methods of `CloudProvider`.
+func PublishLatency(cloudProvider cloudprovider.CloudProvider) cloudprovider.CloudProvider {
+	switch c := cloudProvider.(type) {
```

Comment:
Why bother checking whether or not you're passing in a decorator? Easier to just do the default in all cases and then never call this twice in a row.

Reply:
Double-decoration would be a mistake (each method call would be counted twice) but would not produce a compile-time or runtime error.  Karpenter would simply publish incorrect metric values.

Detecting and fixing this situation is easy, so it was added to the code.  The alternative would be to return an `error`.

Reply:
There are an infinite number of programming mistakes that could be made that would result in undesirable behavior. Attempting to protect against them with more code increases the surface area of the project and needs to be understood, tested. 

This code base follows a pattern where our external interfaces (e.g. provisioner) are validated and protected, but for internal interfaces, we assume that developers are doing the right thing. This allows the code to accurately reflect its purpose, without the additional complexity of whether or not a developer decided to protect against a specific failure mode.